state/export_test.go

```diff
@@ -183,31 +184,75 @@ func AddTestingCharmMultiSeries(c *gc.C, st *State, name string) *Charm {
 }
 
 func AddTestingApplication(c *gc.C, st *State, name string, ch *Charm) *Application {
-	return addTestingApplication(c, st, "", name, ch, nil, nil)
+	return addTestingApplication(addTestingApplicationParams{
+		c:    c,
+		st:   st,
+		name: name,
+		ch:   ch,
+	})
 }
 
 func AddTestingApplicationForSeries(c *gc.C, st *State, series, name string, ch *Charm) *Application {
-	return addTestingApplication(c, st, series, name, ch, nil, nil)
+	return addTestingApplication(addTestingApplicationParams{
+		c:      c,
+		st:     st,
+		series: series,
+		name:   name,
+		ch:     ch,
+	})
 }
 
 func AddTestingApplicationWithStorage(c *gc.C, st *State, name string, ch *Charm, storage map[string]StorageConstraints) *Application {
-	return addTestingApplication(c, st, "", name, ch, nil, storage)
+	return addTestingApplication(addTestingApplicationParams{
+		c:       c,
+		st:      st,
+		name:    name,
+		ch:      ch,
+		storage: storage,
+	})
+}
+
+func AddTestingApplicationWithDevices(c *gc.C, st *State, name string, ch *Charm, devices map[string]DeviceConstraints) *Application {
+	return addTestingApplication(addTestingApplicationParams{
+		c:       c,
+		st:      st,
+		name:    name,
+		ch:      ch,
+		devices: devices,
+	})
 }
 
 func AddTestingApplicationWithBindings(c *gc.C, st *State, name string, ch *Charm, bindings map[string]string) *Application {
-	return addTestingApplication(c, st, "", name, ch, bindings, nil)
+	return addTestingApplication(addTestingApplicationParams{
+		c:        c,
+		st:       st,
+		name:     name,
+		ch:       ch,
+		bindings: bindings,
+	})
 }
 
-func addTestingApplication(c *gc.C, st *State, series, name string, ch *Charm, bindings map[string]string, storage map[string]StorageConstraints) *Application {
-	c.Assert(ch, gc.NotNil)
-	app, err := st.AddApplication(AddApplicationArgs{
-		Name:             name,
-		Series:           series,
-		Charm:            ch,
-		EndpointBindings: bindings,
-		Storage:          storage,
+type addTestingApplicationParams struct {
+	c            *gc.C
+	st           *State
+	series, name string
+	ch           *Charm
+	bindings     map[string]string
+	storage      map[string]StorageConstraints
+	devices      map[string]DeviceConstraints
+}
+
+func addTestingApplication(params addTestingApplicationParams) *Application {
```

Comment:
We always tend to pass in gc.C as the first parameter.
Don't add it as a params attribute.